packages/server/src/runner.ts

```diff
@@ -0,0 +1,170 @@
+import { Operation } from 'effection';
+import { Test, TestResult } from '@bigtest/suite';
+import { Mailbox } from '@bigtest/effection';
+import { Atom } from '@bigtest/atom';
+import { AgentState, OrchestratorState, BundlerState } from './orchestrator/state';
+import { TestRunAggregator } from './result-aggregator/test-run';
+import { filterTest } from './filter-test';
+import { SpawnContext } from './spawn-context';
+import { resultStream } from './result-stream';
+import { TestEvent } from './schema/test-event';
+
+import { ConnectionChannel } from './connection-server';
+
+export interface RunnerOptions {
+  context: SpawnContext;
+  atom: Atom<OrchestratorState>;
+  agents: ConnectionChannel;
+  proxyPort: number;
+  manifestPort: number;
+  testFiles?: string[];
+};
+
+export interface RunOptions {
+  testRunId: string;
+  files: string[];
+}
+
+export interface Runner {
+  run(options: RunOptions): PromiseLike<void>;
+  subscribe(id: string): AsyncIterator<TestEvent>;
+}
+
+export class AgentRunner implements Runner {
+  constructor(public options: RunnerOptions) {
+  }
+
+  run(options: RunOptions): PromiseLike<void> {
+    return this.options.context.spawn(this._run(options));
+  };
+
+  async *subscribe(id: string): AsyncIterator<TestEvent> {
+    let slice = this.options.atom.slice('testRuns', id);
+
+    let scope = this.options.context.spawn(undefined) as SpawnContext;
+
+    let subscription = await scope.spawn(resultStream(id, slice))
+
+    try {
+      while(true) {
+        let iter = await scope.spawn(subscription.next());
+        if(iter.done) {
+          break;
+        } else {
+          yield iter.value;
+        }
+      }
+    } finally {
+      scope.halt();
+    }
+  }
+
+  private *_run({ testRunId, files }: RunOptions): Operation<void> {
+    console.debug('[command processor] running test', testRunId);
+    let stepTimeout = 60_000;
+    let testRunSlice = this.options.atom.slice('testRuns', testRunId);
+
+    let bundlerSlice = this.options.atom.slice('bundler');
+
+    let bundler: BundlerState = yield bundlerSlice.once((state) => state.type === 'GREEN' || state.type === 'ERRORED');
+
+    if(bundler.type === 'GREEN') {
+      let events = yield Mailbox.from(this.options.agents.match({ testRunId }));
```

Comment:
@cowboyd I would love to hear your thoughts on this pattern.

The reasoning behind this is roughly to use mailboxes as sort of a supercharged subscription. Unlike `Subscription`, a mailbox can selectively pull certain messages, while leaving the remaining messages in the inbox. However, there is a problem we've seen with mailboxes where creating a global mailbox makes it very clear who pulls from it and when, and how any remaining messages get cleaned up. Here I've tried using channels (or in this case a DuplexChannel) and mailboxes *together*. We're creating a mailbox specifically for this test run, and once it is done, we no longer need it and can throw it away.

Reply:
I think this is a good pattern in general, and I was thinking something similar on my [comment about a queue primitive][1]. I think this pattern will be necessary to avoid race conditions, but also to avoid having shared clobber-prone spaces. Eventually, I'd like to see mailbox evolve into a general buffer class that can be used as a basis for lifo and fifo queues, or any other number of stateful structures.

[1]: https://github.com/thefrontside/effection/issues/212#issuecomment-726495079